lib/debug-log.js

```diff
@@ -0,0 +1,174 @@
+  /** @type {RokuDebugLogOpts} */
+  #opts;
+
+  /** @type {string} */
+  #host;
+
+  /**
+   * @type {string}
+   */
+  #hostSlug;
+
+  /** @type {string} */
+  #logDirpath;
+  /** @type {string} */
+  #tempDirpath;
+
+  /**
+   *
+   * @param {string} deviceHost
+   * @param {RokuDebugLogOpts} [opts]
+   */
+  constructor(deviceHost, opts = /** @type {RokuDebugLogOpts} */ ({})) {
+    this.#host = deviceHost;
+    this.#opts = _.defaultsDeep(opts, DEFAULT_OPTS);
+
+    this.#hostSlug = slug(deviceHost);
+
+    const {log: logDirpath, temp: tempDirpath} = envPaths('appium-roku-driver');
+    this.#logDirpath = logDirpath;
+    this.#tempDirpath = tempDirpath;
+
+    if (!this.#opts.logPath) {
+      const logFileBasename = `debug-${this.#hostSlug}.log`;
+      this.#opts.logPath = path.join(logDirpath, logFileBasename);
+    }
+  }
+
+  get logPath() {
+    return this.#opts.logPath;
+  }
+
+  /**
+   *
+   * @param {string} basename
+   * @param {string} logPath
+   * @param {string} tempDirpath
+   * @param {LockLogfileOpts} [opts]
+   * @returns {Promise<() => Promise<void>>}
+   */
+  async #lockLogfile(basename, logPath, tempDirpath, opts = {}) {
+    // `prepare-lockfile` does not support file descriptors
+    if (!_.isNil(opts.fd)) {
+      const lockfilePath = path.join(tempDirpath, `${basename}.lock`);
+      const originalUnlock = await lock(logPath, {...BASE_LOCK_OPTS, lockfilePath});
+      const wrappedUnlock = async () => {
+        try {
+          await originalUnlock();
+          log.debug(`Unlocked ${logPath}`);
+        } catch (err) {
+          // todo handle better
+          log.warn(err);
+        }
+      };
+      return wrappedUnlock;
+    }
+    return () => B.resolve();
+  }
+
+  /**
+   * Prepares filesystem for writing debug log
+   */
+  async #prepareFs() {
+    await B.all([fs.mkdirp(this.#logDirpath), fs.mkdirp(this.#tempDirpath)]);
+  }
+
+  /**
+   * Connects to a debug log and pipes its output somewhere.  Returns a wrapper around the socket.
+   * @returns {Promise<import('telnet-client').Telnet>} - Telnet client
+   */
+  async connect() {
+    await this.#prepareFs();
+
+    const {logPath, logFd} = this.#opts;
+
+    const unlock = await this.#lockLogfile(this.#hostSlug, this.logPath, this.#tempDirpath, {
+      fd: logFd,
+    });
+    const client = new Telnet();
+    await client.connect({
+      host: this.#host,
+      port: this.#opts.port,
+      ...this.#opts.telnetOpts,
+    });
+    log.debug(`Connected to ${this.#host}`);
+
+    // TODO: expose this stream?
+    client
+      .getSocket()
+      // note: the encoding likely to be just `ascii`
+      .pipe(_fs.createWriteStream(logPath, {fd: logFd, encoding: 'utf8'}))
+      .on('error', (err) => {
+        unlock();
+        log.debug(`error: ${err}`);
+      })
+      .on('finish', () => {
+        // todo delete it
+        log.info(`Finished writing debug log to ${logPath}`);
+        unlock();
+      });
+
+    return client;
```

Comment:
this API could be better.  probably should just...not return the `Telnet` instance and keep it somewhere in the class instead, and create wrapper methods as needed.

Reply:
okay, now it wraps the `Telnet` instance instead.